File: src/main/java/gargoyle/sexbomb/game/beans/Game.java
package gargoyle.sexbomb.game.beans;

import gargoyle.sexbomb.game.event.GameEventSource;
import gargoyle.sexbomb.game.event.StatusChangedGameEvent;

import javax.swing.ImageIcon;
import java.awt.Image;
import java.io.Serializable;

public class Game implements Serializable {
    private static final String STR_STATUS__S_S_D = "%s | %s | Left: %d";
    public final GameEventSource source;
    private Campaign campaign;
    private transient Image cover;
    private Field field;
    private transient Image image;
    private Level level;
    private boolean resize;
    private Skin skin;
    private Status status;

    public Game() {
        status = null;
        field = null;
        image = null;
        cover = null;
        level = null;
        skin = null;
        campaign = null;
        source = new GameEventSource();
    }

    public Campaign getCampaign() {
        return campaign;
    }

    public Image getCover() {
        return cover;
    }

    public Field getField() {
        return field;
    }

    public Image getImage() {
        return image;
    }

    public Level getLevel() {
        return level;
    }

    public Skin getSkin() {
        return skin;
    }

    public void setSkin(Skin skin) {
        this.skin = skin;
    }

    public Status getStatus() {
        return status;
    }

    public String getStatusString() {
        return String.format(STR_STATUS__S_S_D, campaign.getName(), level.getName(), field.getLeft() - field.getMines());
    }

    public boolean isGame() {
        return field != null && status == Status.GAME;
    }

    public boolean isLost() {
        return field != null && status == Status.LOST;
    }

    public boolean isResize() {
        return resize;
    }

    public void setResize(boolean resize) {
        this.resize = resize;
    }

    public boolean isWonGame() {
        return field != null && status == Status.WON_GAME;
    }

    public boolean isWonLevel() {
        return field != null && status == Status.WON_LEVEL;
    }

    public void lost() {
        status = Status.LOST;
        source.fireGameEvent(new StatusChangedGameEvent(this, Status.LOST));
    }

    public boolean next() {
        level = campaign.getNextLevel(level);
        if (level == null) {
            image = null;
            status = Status.WON_GAME;
            source.fireGameEvent(new StatusChangedGameEvent(this, Status.WON_GAME));
            return false;
        }
        field.init(level.getWidth(), level.getHeight(), level.getMines());
        if (level.getImage() != null) {
            image = new ImageIcon(level.getImage()).getImage();
        }
        if (level.getCover() != null) {
            cover = new ImageIcon(level.getCover()).getImage();
        }
        status = Status.GAME;
        source.fireGameEvent(new StatusChangedGameEvent(this, Status.GAME));
        return true;
    }

    public void reset() {
        status = null;
    }

    public void start() {
        if (field == null) {
            field = new Field();
        }
        level = null;
        next();
    }

    public void start(Campaign campaign) {
        this.campaign = campaign;
        start();
    }

    public void wonGame() {
        status = Status.WON_GAME;
        source.fireGameEvent(new StatusChangedGameEvent(this, Status.WON_GAME));
    }

    public void wonLevel() {
        status = Status.WON_LEVEL;
        source.fireGameEvent(new StatusChangedGameEvent(this, Status.WON_LEVEL));
    }
}

File: src/main/kotlin/gargoyle/sexbomb/game/beans/Game.kt
package gargoyle.sexbomb.game.beans

import gargoyle.sexbomb.game.event.GameEventSource
import gargoyle.sexbomb.game.event.StatusChangedGameEvent
import java.awt.Image
import java.io.Serializable
import javax.swing.ImageIcon

class Game : Serializable {
    val source: GameEventSource = GameEventSource()
    lateinit var campaign: Campaign
        private set

    lateinit var cover: Image
        private set
    lateinit var field: Field
        private set

    fun fieldInitialized() = ::field.isInitialized

    var image: Image? = null
        private set
    var level: Level? = null
        private set
    var isResize = false
    lateinit var skin: Skin
    var status: Status = Status.NULL
        private set

    fun skinInitialized() = ::skin.isInitialized

    val statusString: String
        get() = String.format(
            STR_STATUS__S_S_D,
            campaign.name,
            level?.name,
            this.field.left - this.field.mines
        )
    val isGame: Boolean
        get() = status == Status.GAME
    val isLost: Boolean
        get() = status == Status.LOST
    val isWonGame: Boolean
        get() = status == Status.WON_GAME
    val isWonLevel: Boolean
        get() = status == Status.WON_LEVEL

    fun lost() {
        status = Status.LOST
        source.fireGameEvent(StatusChangedGameEvent(this, Status.LOST))
    }

    operator fun next(): Boolean {
        level = campaign.getNextLevel(level)
        if (level == null) {
            image = null
            status = Status.WON_GAME
            source.fireGameEvent(StatusChangedGameEvent(this, Status.WON_GAME))
            return false
        }
        field.init(level!!.width, level!!.height, level!!.mines)
        if (level!!.image != null) {
            image = ImageIcon(level!!.image).image
        }
        if (level!!.cover != null) {
            cover = ImageIcon(level!!.cover).image
        }
        status = Status.GAME
        source.fireGameEvent(StatusChangedGameEvent(this, Status.GAME))
        return true
    }

    fun reset() {
        status = Status.NULL
    }

    fun start() {
        if (!::field.isInitialized) {
            field = Field()
        }
        level = null
        next()
    }

    fun start(campaign: Campaign) {
        this.campaign = campaign
        start()
    }

    fun wonGame() {
        status = Status.WON_GAME
        source.fireGameEvent(StatusChangedGameEvent(this, Status.WON_GAME))
    }

    fun wonLevel() {
        status = Status.WON_LEVEL
        source.fireGameEvent(StatusChangedGameEvent(this, Status.WON_LEVEL))
    }

    companion object {
        private const val STR_STATUS__S_S_D = "%s | %s | Left: %d"
    }
}
